add zmp planner

siyuanfeng-tri opened the pull request:
implement the closed form solution for generating com trajectory from a desired zmp trajectory.
i am still working on the tests, but i'd appreciate some earlier feedback on the actual implementation. @RussTedrake for feature review, thanks!

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4483)
<!-- Reviewable:end -->

hongkai-dai commented:
Checkpoint.

---

Reviewed 1 of 10 files at r1.
Review status: 1 of 10 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/ZMP/zmp_test_util.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KYw4JrE-zv6_Btp__Nm:-KYw4JrE-zv6_Btp__Nn:b-9lot20) ([raw file](https://github.com/robotlocomotion/drake/blob/6d2b3b7a2354348fdb3c1d71356a15edb0fcbc60/drake/examples/ZMP/zmp_test_util.h#L13)):*
> ```objc
> // the policy from a drake::systems::ZMPPlanner.
> struct ZMPTestTraj {
>   Eigen::Matrix<double, 1, Eigen::Dynamic> time;
> ```

Use `Eigen::RowVectorXd` for brevity?

---

*[drake/examples/ZMP/zmp_test_util.h, line 37 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KYw4yGWPR4FSmqnfVo1:-KYw4yGWPR4FSmqnfVo2:b-d1h9xm) ([raw file](https://github.com/robotlocomotion/drake/blob/6d2b3b7a2354348fdb3c1d71356a15edb0fcbc60/drake/examples/ZMP/zmp_test_util.h#L37)):*
> ```objc
>  * value function, and linear policy.
>  * @param x0, Initial condition.
>  * @param dt, Time step
> ```

. at the end

---

*[drake/examples/ZMP/zmp_test_util.h, line 51 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KYw5Fm5mLWkfa6Ok-L5:-KYw5Fm5mLWkfa6Ok-L6:bhpftr6) ([raw file](https://github.com/robotlocomotion/drake/blob/6d2b3b7a2354348fdb3c1d71356a15edb0fcbc60/drake/examples/ZMP/zmp_test_util.h#L51)):*
> ```objc
> /**
>  * Generates desired ZMP trajectories as piecewise polynomials given the
>  * desired footsteps. The knot points are generated as follows:
> ```

Could you explain here `ss` stands for `single support`, `ds` stands for `double support`, `fs` stands for `foot step`?

---

*[drake/examples/ZMP/zmp_test_util.h, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KYw5vzgPT6QbxShLosi:-KYw5vzgPT6QbxShLosj:b67fzql) ([raw file](https://github.com/robotlocomotion/drake/blob/6d2b3b7a2354348fdb3c1d71356a15edb0fcbc60/drake/examples/ZMP/zmp_test_util.h#L62)):*
> ```objc
>  * @param double_support_duration, Duration for double support.
>  * @param single_support_duration, Duration for single support.
>  * @return Three trajectories, 0 is piecewise constant, 1 is linear, 2 is cubic.
> ```

maybe use `zero-order-hold` instead of `constant`, since `constant` can also mean `backward zero-order-hold`, namely  `x(t) = x[i] if t[i-1] <= t <= t[i]`.

Why we need a piecewise constant desired ZMP trajectory? The resultant control (the CoM acceleration) will be discontinuous when the desired ZMP jumps from one location to the next location.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-:-KYw6egweKh7I_39gWfX:by96gdd)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 1 of 10 files at r1.
Review status: 2 of 10 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/examples/ZMP/zmp_test_util.cc, line 15 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KYw9En3xeGw6_AVpWRx:-KYw9En3xeGw6_AVpWRy:b-d8qsp4) ([raw file](https://github.com/robotlocomotion/drake/blob/6d2b3b7a2354348fdb3c1d71356a15edb0fcbc60/drake/examples/ZMP/zmp_test_util.cc#L15)):*
> ```cpp
>       (zmp_d.getEndTime() + T_final_offset - zmp_d.getStartTime()) / dt);
> 
>   ZMPTestTraj traj(N);
> ```

The value of `traj` is not initialized, so what are we simulating here?

---

*[drake/examples/ZMP/zmp_test_util.h, line 33 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KYw9XCTtDa0XhUn6vaJ:-KYw9XCTtDa0XhUn6vaK:b-mr0bgz) ([raw file](https://github.com/robotlocomotion/drake/blob/6d2b3b7a2354348fdb3c1d71356a15edb0fcbc60/drake/examples/ZMP/zmp_test_util.h#L33)):*
> ```objc
> /**
>  * Forward simulation using the linear policy from `zmp_planner` starting from
>  * the initial condition `x0` using Euler integration.
> ```

Could you say it is `explicit Euler integration`, to differentiate it from `implicit Euler integration` (`x[n+1] = x[n] + f(x[n+1], u[n+1])dt`)?

Actually, why we are writing an Euler integration here, instead creating a simple linear system, and then call the existing simulators for this dynamical system?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4483)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
+@hongkai-dai for feature review.

---

Review status: 2 of 10 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-:-KYyP85Rsn7eXsUKA4Re:bl5ccwd)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: 2 of 10 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/examples/ZMP/zmp_test_util.cc, line 15 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KYw9En3xeGw6_AVpWRx:-KYyXbteUV2AMGKkse4d:b7rwkne) ([raw file](https://github.com/robotlocomotion/drake/blob/6d2b3b7a2354348fdb3c1d71356a15edb0fcbc60/drake/examples/ZMP/zmp_test_util.cc#L15)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

The value of `traj` is not initialized, so what are we simulating here?
</blockquote></details>

Oh, they were filled down in the next for loop block.

---

*[drake/examples/ZMP/zmp_test_util.h, line 13 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KYw4JrE-zv6_Btp__Nm:-KYyY4n1pFC7UX5Kbdho:b-stg8os) ([raw file](https://github.com/robotlocomotion/drake/blob/6d2b3b7a2354348fdb3c1d71356a15edb0fcbc60/drake/examples/ZMP/zmp_test_util.h#L13)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Use `Eigen::RowVectorXd` for brevity?
</blockquote></details>

i actually like it better this way, so it aligns better with the rest. 

---

*[drake/examples/ZMP/zmp_test_util.h, line 33 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KYw9XCTtDa0XhUn6vaJ:-KYyYXYdOZn8g3LNiErF:b26gwqk) ([raw file](https://github.com/robotlocomotion/drake/blob/6d2b3b7a2354348fdb3c1d71356a15edb0fcbc60/drake/examples/ZMP/zmp_test_util.h#L33)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Could you say it is `explicit Euler integration`, to differentiate it from `implicit Euler integration` (`x[n+1] = x[n] + f(x[n+1], u[n+1])dt`)?

Actually, why we are writing an Euler integration here, instead creating a simple linear system, and then call the existing simulators for this dynamical system?
</blockquote></details>

i wanted this to be a barebone test, and this is the easiest way to do that.
i could write a separate system test / example, but that would be testing the system not really for the zmp planner. 

---

*[drake/examples/ZMP/zmp_test_util.h, line 37 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KYw4yGWPR4FSmqnfVo1:-KYyZ5o-v7IW53Wg88FD:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6d2b3b7a2354348fdb3c1d71356a15edb0fcbc60/drake/examples/ZMP/zmp_test_util.h#L37)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

. at the end
</blockquote></details>

Done.

---

*[drake/examples/ZMP/zmp_test_util.h, line 51 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KYw5Fm5mLWkfa6Ok-L5:-KYyZ6fRegq9IIaJ_Di3:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6d2b3b7a2354348fdb3c1d71356a15edb0fcbc60/drake/examples/ZMP/zmp_test_util.h#L51)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Could you explain here `ss` stands for `single support`, `ds` stands for `double support`, `fs` stands for `foot step`?
</blockquote></details>

Done.

---

*[drake/examples/ZMP/zmp_test_util.h, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KYw5vzgPT6QbxShLosi:-KYycXqiO3p2UPRcJQih:bd0b31n) ([raw file](https://github.com/robotlocomotion/drake/blob/6d2b3b7a2354348fdb3c1d71356a15edb0fcbc60/drake/examples/ZMP/zmp_test_util.h#L62)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

maybe use `zero-order-hold` instead of `constant`, since `constant` can also mean `backward zero-order-hold`, namely  `x(t) = x[i] if t[i-1] <= t <= t[i]`.

Why we need a piecewise constant desired ZMP trajectory? The resultant control (the CoM acceleration) will be discontinuous when the desired ZMP jumps from one location to the next location.
</blockquote></details>

we don't actually run it on the robot for the exact same reason you just mentioned. but the zmp planner should still work for a zoh zmp_d. i am just testing that here. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4483)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
I think a possible test is to pretend we do not know that `S1` is a constant, but write down the time derivative of `S1` and `S2` from the Riccati equation. Then integrate `S1, S2` backward in time, and then compare to our closed-form solution.

---

Review status: 1 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/ZMP/test/zmp_planner_test.cc, line 16 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KYzWeeb9WpjuB-9ymwM:-KYzWeecnEMQv5i80e7n:bj4omyf) ([raw file](https://github.com/robotlocomotion/drake/blob/6d2b3b7a2354348fdb3c1d71356a15edb0fcbc60/drake/examples/ZMP/test/zmp_planner_test.cc#L16)):*
> ```cpp
> // where x_bar = [com - zmp_d_tf; comd],
> // the optimal control u (CoM acceleration) is achieved by:
> // min_u L(y, u) + dV / dx_bar * x_bar_dot,
> ```

Could you add a check for HJB equation here, that 
```
0 = x_bar^T*S2_dot + min_u L(y, u) + dV /dx_bar * x_bar_dot
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-:-KYzh5RdFSNkFQ_CvNvx:b4x0v1u)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 2 of 10 files at r1.
Review status: 3 of 10 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/systems/controllers/zmp_planner.cc, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZ2UjClwQ0ysvMvO43m:-KZ2UjClwQ0ysvMvO43n:by2s2em) ([raw file](https://github.com/robotlocomotion/drake/blob/66541ab2d81ddd8767cc2b14947e5d8a22ddda16/drake/systems/controllers/zmp_planner.cc#L23)):*
> ```cpp
>                       const Eigen::Vector4d& x0, double height,
>                       const Eigen::Matrix2d& Qy, const Eigen::Matrix2d& R) {
>   int n_segments = zmp_d.getNumberOfSegments();
> ```

Assert `height > 0 `.

---

*[drake/systems/controllers/zmp_planner.cc, line 41 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZ2UWlLP2W9PetMzOvv:-KZ2UWlLP2W9PetMzOvw:boo3qv1) ([raw file](https://github.com/robotlocomotion/drake/blob/66541ab2d81ddd8767cc2b14947e5d8a22ddda16/drake/systems/controllers/zmp_planner.cc#L41)):*
> ```cpp
>   C_.setZero();
>   C_.block<2, 2>(0, 0).setIdentity();
>   D_ = -height / 9.81 * Eigen::Matrix2d::Identity();
> ```

Maybe use a variable `gravity` in the class to store the value, and we can change the value in the future.

---

*[drake/systems/controllers/zmp_planner.cc, line 68 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZ2WkRtBRSEFPWrnz8d:-KZ2WkRtBRSEFPWrnz8e:bvbs410) ([raw file](https://github.com/robotlocomotion/drake/blob/66541ab2d81ddd8767cc2b14947e5d8a22ddda16/drake/systems/controllers/zmp_planner.cc#L68)):*
> ```cpp
>   Eigen::Vector4d tmp4;
> 
>   Eigen::MatrixXd alpha = Eigen::MatrixXd::Zero(4, n_segments);
> ```

To make `alpha` look more similar to `beta` and `gamma`, maybe use `std::vector<Vector4d>` for `alpha`?

---

*[drake/systems/controllers/zmp_planner.cc, line 155 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZ2Z41ttRSbstwHVN3p:-KZ2Z41uXa5ilf2i-GnQ:b5frh14) ([raw file](https://github.com/robotlocomotion/drake/blob/66541ab2d81ddd8767cc2b14947e5d8a22ddda16/drake/systems/controllers/zmp_planner.cc#L155)):*
> ```cpp
> 
>   Eigen::MatrixXd a(8, n_segments);
>   a.bottomRows(4) = alpha;
> ```

a.bottomsRows<4>()

---

*[drake/systems/controllers/zmp_planner.h, line 53 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZ2Mg0QrPBTg2nUsqn4:-KZ2Mg0QrPBTg2nUsqn5:b-bnfx22) ([raw file](https://github.com/robotlocomotion/drake/blob/66541ab2d81ddd8767cc2b14947e5d8a22ddda16/drake/systems/controllers/zmp_planner.h#L53)):*
> ```objc
>   /**
>    * Converts CoM acceleration to center of pressure (CoP) using
>    *    cop = C * x + D * u
> ```

If it helps, you may also write down the equation to compute cop from linear inverted pendulum model, that
```
cop_x = com_x - com_z/g * comddot_x
cop_y = com_y - com_z/g * comddot_x
```

---

*[drake/systems/controllers/zmp_planner.h, line 173 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZ2RCcGt050ifyV5Fqh:-KZ2RCcGt050ifyV5Fqi:bgqf4ty) ([raw file](https://github.com/robotlocomotion/drake/blob/66541ab2d81ddd8767cc2b14947e5d8a22ddda16/drake/systems/controllers/zmp_planner.h#L173)):*
> ```objc
>    * Returns the time invariant second order term of the value function.
>    */
>   inline const Eigen::Matrix<double, 4, 4>&
> ```

Eigen::Matrix4d.

---

*[drake/systems/controllers/zmp_planner.h, line 207 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZ2TcCbVakphwRfFyhN:-KZ2TcCbVakphwRfFyhO:byst9cq) ([raw file](https://github.com/robotlocomotion/drake/blob/66541ab2d81ddd8767cc2b14947e5d8a22ddda16/drake/systems/controllers/zmp_planner.h#L207)):*
> ```objc
>   // Desired ZMP and ZMPd trajectories.
>   PiecewisePolynomial<double> zmp_d_;
>   PiecewisePolynomial<double> zmpd_d_;
> ```

Why we need desired zmpd trajectory?

---

*[drake/systems/controllers/zmp_planner.h, line 215 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZ2TheO7qW9pvn2iUjT:-KZ2TheO7qW9pvn2iUjU:bicjpj0) ([raw file](https://github.com/robotlocomotion/drake/blob/66541ab2d81ddd8767cc2b14947e5d8a22ddda16/drake/systems/controllers/zmp_planner.h#L215)):*
> ```objc
> 
>   // System dynamics matrices.
>   Eigen::Matrix<double, 4, 4> A_;
> ```

const matrix?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4483)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 1 of 10 files at r1.
Review status: 4 of 10 files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4483)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Since `s2` is an exponential plus trajectory, could you take the time derivative of `s2`, and then compare with Equation 18 in the paper?

---

Reviewed 1 of 10 files at r1, 2 of 2 files at r2.
Review status: 7 of 10 files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-:-KZ2ym8C07TqnUSL5rzh:bvale0e)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
yeah, that's my plan, it seems the easiest to check.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-:-KZ2ziXTjvtuxALxVDPz:bkfsuh0)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
@drake-jenkins-bot linux-gcc-bazel-experimental please
@drake-jenkins-bot linux-xenial-gcc-bazel-experimental please

siyuanfeng-tri commented:
Review status: 1 of 10 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/examples/ZMP/test/zmp_planner_test.cc, line 16 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KYzWeeb9WpjuB-9ymwM:-KZ9-b0Y6jzkA10v_OPI:b-b97aov) ([raw file](https://github.com/robotlocomotion/drake/blob/6d2b3b7a2354348fdb3c1d71356a15edb0fcbc60/drake/examples/ZMP/test/zmp_planner_test.cc#L16)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Could you add a check for HJB equation here, that 
```
0 = x_bar^T*S2_dot + min_u L(y, u) + dV /dx_bar * x_bar_dot
```
</blockquote></details>

see new tests. 

---

*[drake/systems/controllers/zmp_planner.cc, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZ2UjClwQ0ysvMvO43m:-KZ2iWf0I-L_xI1MvW6i:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/66541ab2d81ddd8767cc2b14947e5d8a22ddda16/drake/systems/controllers/zmp_planner.cc#L23)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Assert `height > 0 `.
</blockquote></details>

Done.

---

*[drake/systems/controllers/zmp_planner.cc, line 41 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZ2UWlLP2W9PetMzOvv:-KZ2jBBFNwi_0BUL0Yh1:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/66541ab2d81ddd8767cc2b14947e5d8a22ddda16/drake/systems/controllers/zmp_planner.cc#L41)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Maybe use a variable `gravity` in the class to store the value, and we can change the value in the future.
</blockquote></details>

Done.

---

*[drake/systems/controllers/zmp_planner.cc, line 68 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZ2WkRtBRSEFPWrnz8d:-KZ2juX0NngYrToAwpXZ:blhoyie) ([raw file](https://github.com/robotlocomotion/drake/blob/66541ab2d81ddd8767cc2b14947e5d8a22ddda16/drake/systems/controllers/zmp_planner.cc#L68)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

To make `alpha` look more similar to `beta` and `gamma`, maybe use `std::vector<Vector4d>` for `alpha`?
</blockquote></details>

alpha beta, and gamma are consumed by different constructors for polynomial.. i think it needs to look like this actually. 

---

*[drake/systems/controllers/zmp_planner.cc, line 155 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZ2Z41ttRSbstwHVN3p:-KZ2k6E-zr8pPni1mkx7:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/66541ab2d81ddd8767cc2b14947e5d8a22ddda16/drake/systems/controllers/zmp_planner.cc#L155)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

a.bottomsRows<4>()
</blockquote></details>

Done.

---

*[drake/systems/controllers/zmp_planner.h, line 53 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZ2Mg0QrPBTg2nUsqn4:-KZ9034nm2cFAz6lmcTE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/66541ab2d81ddd8767cc2b14947e5d8a22ddda16/drake/systems/controllers/zmp_planner.h#L53)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

If it helps, you may also write down the equation to compute cop from linear inverted pendulum model, that
```
cop_x = com_x - com_z/g * comddot_x
cop_y = com_y - com_z/g * comddot_x
```
</blockquote></details>

Done.

---

*[drake/systems/controllers/zmp_planner.h, line 173 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZ2RCcGt050ifyV5Fqh:-KZ2kq1TDqs1fwDH5NZ6:bq5qkw5) ([raw file](https://github.com/robotlocomotion/drake/blob/66541ab2d81ddd8767cc2b14947e5d8a22ddda16/drake/systems/controllers/zmp_planner.h#L173)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Eigen::Matrix4d.
</blockquote></details>

i like it better this way, more consistent with the abcd stuff. 

---

*[drake/systems/controllers/zmp_planner.h, line 207 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZ2TcCbVakphwRfFyhN:-KZ2_PmdcDqkURRnGltd:b-mc2edk) ([raw file](https://github.com/robotlocomotion/drake/blob/66541ab2d81ddd8767cc2b14947e5d8a22ddda16/drake/systems/controllers/zmp_planner.h#L207)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Why we need desired zmpd trajectory?
</blockquote></details>

not really. i should probably take it out. i vaguely remembered using it for debugging. 

---

*[drake/systems/controllers/zmp_planner.h, line 215 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZ2TheO7qW9pvn2iUjT:-KZ9086-nsBIHCIkHz0E:b-11b7mi) ([raw file](https://github.com/robotlocomotion/drake/blob/66541ab2d81ddd8767cc2b14947e5d8a22ddda16/drake/systems/controllers/zmp_planner.h#L215)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

const matrix?
</blockquote></details>

it's a bit easier to read if i init these in the constructor body instead of the initialization list. so i am inclined to leave them as is. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4483)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 1 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/controllers/zmp_planner.h, line 161 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZD52PlKnKy8SH_AC1g:-KZD52PlKnKy8SH_AC1h:b-es6ahl) ([raw file](https://github.com/robotlocomotion/drake/blob/3c995208bb11adeecb5f873828df6fe2e92045cf/drake/systems/controllers/zmp_planner.h#L161)):*
> ```objc
>    * Returns the time invariant second order term of the value function.
>    */
>   inline const Eigen::Matrix<double, 4, 4>& get_Vxx() const {
> ```

If you use value function `V`, could you write out the explicit form of value function as
`V = xbar*S1*xbar + xbar*s2(t)+s3` in the documentation, so that people know the value function is the optimal cost-to-go function? In Russ's paper I think he did not explicitly say that is the value function, and people may not know what `value function` refers to by just reading the paper.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-:-KZD5h_CXIDfSMCUe_dd:bnfp4nl)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: 1 of 11 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/controllers/zmp_planner.h, line 161 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZD52PlKnKy8SH_AC1g:-KZMjm9jz_CtElNDryQd:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3c995208bb11adeecb5f873828df6fe2e92045cf/drake/systems/controllers/zmp_planner.h#L161)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

If you use value function `V`, could you write out the explicit form of value function as
`V = xbar*S1*xbar + xbar*s2(t)+s3` in the documentation, so that people know the value function is the optimal cost-to-go function? In Russ's paper I think he did not explicitly say that is the value function, and people may not know what `value function` refers to by just reading the paper.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4483)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
@sherm1 for platform review. thanks much!

siyuanfeng-tri commented:
*[drake/systems/controllers/zmp_planner.cc, line 42 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZMkRODgPxt6-nwVEQJ:-KZMkRODgPxt6-nwVEQK:bdwc5t9) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/systems/controllers/zmp_planner.cc#L42)):*
> ```cpp
>   if (!CheckStationaryEndPoint(zmp_d)) {
>     std::cerr << "ZMPPlanner: The desired zmp trajectory does not end in a "
>                  "stationary condition.\n";
> ```

I am not sure how to do this part properly. I sort of just want to dump a warning message, any suggestions ? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4483)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
@drake-jenkins-bot linux-gcc-experimental-ros please

sherm1 commented:
Platform review pass done, PTAL. Looks very nice, but I have no idea if it is correct -- I hope it was well feature reviewed!

---

Reviewed 2 of 10 files at r1, 1 of 4 files at r3, 3 of 6 files at r4, 5 of 5 files at r5.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/examples/ZMP/zmp_test_util.cc, line 49 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZOKtp5zcarNo_PeQYZ:-KZOKtp6EyRvohgqby1T:b-bis96o) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/examples/ZMP/zmp_test_util.cc#L49)):*
> ```cpp
>   DRAKE_DEMAND(!footsteps.empty());
> 
>   std::vector<double> Ts;
> ```

Something less obscure and style guide compliant, maybe `time_steps`?

---

*[drake/examples/ZMP/zmp_test_util.h, line 11 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZOIXtAZWYprcOjgozU:-KZOIXtAZWYprcOjgozV:b-33kwwb) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/examples/ZMP/zmp_test_util.h#L11)):*
> ```objc
> 
> // A structure for storing trajectories from simulating a linear inverted
> // pendulum model (LIPM) using the policy from a drake::systems::ZMPPlanner.
> ```

This is visible so should have a doxygen comment. Or you could put it in a local namespace like `zmp::internal` or `zmp::detail`.

---

*[drake/examples/ZMP/zmp_test_util.h, line 21 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZOJwPcjnsw7F3NTPgm:-KZOJwPcjnsw7F3NTPgn:bvmfzv5) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/examples/ZMP/zmp_test_util.h#L21)):*
> ```objc
>   Eigen::Matrix<double, 2, Eigen::Dynamic> cop;
> 
>   explicit ZMPTestTraj(int N) {
> ```

[Style guide](https://google.github.io/styleguide/cppguide.html#Declaration_Order) wants method before data members.

---

*[drake/examples/ZMP/zmp_test_util.h, line 38 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZOKTWBM7jafoYW4C57:-KZOKTWBM7jafoYW4C58:bxv5nce) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/examples/ZMP/zmp_test_util.h#L38)):*
> ```objc
>  * @param x0, Initial condition.
>  * @param dt, Time step.
>  * @param T_final_offset, Simulate until `T_final_offset` pass the end of the
> ```

Use lower case `t_final_offset` per style guide. (Also we use "T" everywhere as the template scalar type.)

---

*[drake/systems/controllers/zmp_planner.cc, line 27 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZOOd16YI99fxtOomMn:-KZOOd16YI99fxtOomMo:bk2iu57) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/systems/controllers/zmp_planner.cc#L27)):*
> ```cpp
>   int degree = last_segment.getSegmentPolynomialDegree(0);
>   for (int d = degree; d >= 0; d--) {
>     if (derivative.value(derivative.getEndTime()).norm() > 1e-8) return false;
> ```

This `1e-8` appears to be a magic constant. Why does it have this value? Please document.

---

*[drake/systems/controllers/zmp_planner.cc, line 42 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZMkRODgPxt6-nwVEQJ:-KZOEXf-3bgTRWRe47vV:bem4rbb) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/systems/controllers/zmp_planner.cc#L42)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

I am not sure how to do this part properly. I sort of just want to dump a warning message, any suggestions ? 
</blockquote></details>

Yes, you should use the log().

---

*[drake/systems/controllers/zmp_planner.h, line 38 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZOF1LMK7b99liOESJA:-KZOF1LMK7b99liOESJB:b-bkhbkr) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/systems/controllers/zmp_planner.h#L38)):*
> ```objc
>    * @param Qy, Quadratic cost term on ZMP deviation from the desired.
>    * @param R, Quadratic cost term on CoM acceleration.
>    */
> ```

Comment should clarify that this method computes results into internal data members A,B,C,D etc. and that these should be accessed using supplied getters. It would be nice to have a short description of each of the returned quantities -- you already have a lot of that in the private comments.

Also if I'm interpreting this correctly none of the other methods here are valid until after a call to Plan(). You should at least say that but it would be better if the other methods fail if Plan() has not been called rather than return garbage which is what I think they would do now.

---

*[drake/systems/controllers/zmp_planner.h, line 52 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZOFvrQkYUyOGHeMSJp:-KZOFvrQkYUyOGHeMSJq:bsf7og1) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/systems/controllers/zmp_planner.h#L52)):*
> ```objc
>    */
>   Eigen::Vector2d ComputeOptimalCoMdd(double time,
>                                       const Eigen::Vector4d& x) const;
> ```

When is it allowed to call this method? Only after a Plan() call? Should document restrictions on when this is valid. Better if you can prevent misuse.

Same comment for all the methods here.

---

*[drake/systems/controllers/zmp_planner.h, line 57 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZOD8wiz5YV9zktRxIL:-KZOD8wiz5YV9zktRxIM:bnjcdoc) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/systems/controllers/zmp_planner.h#L57)):*
> ```objc
>    * Converts CoM acceleration to center of pressure (CoP) using
>    *    cop = C * x + D * u or
>    *    cop = com - z / g * comdd
> ```

BTW, I don't understand the second line here. Are you saying these are the same thing? The code seems just to implement the Cx+Du form. Please clarify.

---

*[drake/systems/controllers/zmp_planner.h, line 62 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZOJ8N_sNVp05IbcsNM:-KZOJ8N_sNVp05IbcsNN:b-gzmfvb) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/systems/controllers/zmp_planner.h#L62)):*
> ```objc
>    * @return center of pressure (CoP)
>    */
>   inline Eigen::Vector2d comdd_to_cop(const Eigen::Vector4d& x,
> ```

Don't put `inline` here. It is implied by putting the implementation in the class and redundant here. The compiler will decide whether the method is actually inlined regardless of your request here. (You do need `inline` if you put the declaration here but put the definition at the end of the header file.)

---

*[drake/systems/controllers/zmp_planner.h, line 68 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZOG_wJqzvXPwLsvMzf:-KZOG_wKxp64oDT88CkY:b-4nnakj) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/systems/controllers/zmp_planner.h#L68)):*
> ```objc
> 
>   /**
>    * Getter for A matrix.
> ```

It would be useful to know when this method can be expected to return a valid result.

---

*[drake/systems/controllers/zmp_planner.h, line 217 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZOHr1mr1W1vGbggfJl:-KZOHr1mr1W1vGbggfJm:byhdkxe) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/systems/controllers/zmp_planner.h#L217)):*
> ```objc
> 
>   // One step cost function:
>   // L = (y - y_d)^T * Qy * (y - y_d)^T + u^T * R * u.
> ```

BTW, this nice comment seems like it would be better placed in the public doxygen (and the ones below too).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-:-KZOP2N892YiQY4WJWhA:b80z9ue)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
added more comments to the class documentation for ZMPPlanner. @hongkai-dai can you take another quick look? thanks. 

---

Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/examples/ZMP/zmp_test_util.cc, line 49 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZOKtp5zcarNo_PeQYZ:-KZP-K5RlTU4YbpP6jfo:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/examples/ZMP/zmp_test_util.cc#L49)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Something less obscure and style guide compliant, maybe `time_steps`?
</blockquote></details>

Done.

---

*[drake/examples/ZMP/zmp_test_util.h, line 11 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZOIXtAZWYprcOjgozU:-KZP-Yqt0mdJaAP6imwv:b-nr7z38) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/examples/ZMP/zmp_test_util.h#L11)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

This is visible so should have a doxygen comment. Or you could put it in a local namespace like `zmp::internal` or `zmp::detail`.
</blockquote></details>

I changed this part to a doxygen comment.

---

*[drake/examples/ZMP/zmp_test_util.h, line 21 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZOJwPcjnsw7F3NTPgm:-KZP-Z7w9ng18g_suwcr:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/examples/ZMP/zmp_test_util.h#L21)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

[Style guide](https://google.github.io/styleguide/cppguide.html#Declaration_Order) wants method before data members.
</blockquote></details>

Done.

---

*[drake/examples/ZMP/zmp_test_util.h, line 38 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZOKTWBM7jafoYW4C57:-KZP05v_EuHj_apn7Gsg:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/examples/ZMP/zmp_test_util.h#L38)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Use lower case `t_final_offset` per style guide. (Also we use "T" everywhere as the template scalar type.)
</blockquote></details>

Done.

---

*[drake/systems/controllers/zmp_planner.cc, line 27 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZOOd16YI99fxtOomMn:-KZP1pSCXhNOJcfUfgBT:bwclwf1) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/systems/controllers/zmp_planner.cc#L27)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

This `1e-8` appears to be a magic constant. Why does it have this value? Please document.
</blockquote></details>

I added a constant in the header file, so it's more obvious. I picked the number rather arbitrarily. 

---

*[drake/systems/controllers/zmp_planner.cc, line 42 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZMkRODgPxt6-nwVEQJ:-KZP2Z4cx1oz3g6VFJwN:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/systems/controllers/zmp_planner.cc#L42)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Yes, you should use the log().
</blockquote></details>

Done.

---

*[drake/systems/controllers/zmp_planner.h, line 38 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZOF1LMK7b99liOESJA:-KZP5ncSoGWHyRnC5jHb:b-dhsmv4) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/systems/controllers/zmp_planner.h#L38)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Comment should clarify that this method computes results into internal data members A,B,C,D etc. and that these should be accessed using supplied getters. It would be nice to have a short description of each of the returned quantities -- you already have a lot of that in the private comments.

Also if I'm interpreting this correctly none of the other methods here are valid until after a call to Plan(). You should at least say that but it would be better if the other methods fail if Plan() has not been called rather than return garbage which is what I think they would do now.
</blockquote></details>

You are correct that all the others are invalid before Plan is called. I added drake demands to all the other public methods.. 

---

*[drake/systems/controllers/zmp_planner.h, line 52 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZOFvrQkYUyOGHeMSJp:-KZP6KpZyYLbyAUotPwR:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/systems/controllers/zmp_planner.h#L52)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

When is it allowed to call this method? Only after a Plan() call? Should document restrictions on when this is valid. Better if you can prevent misuse.

Same comment for all the methods here.
</blockquote></details>

Done.

---

*[drake/systems/controllers/zmp_planner.h, line 57 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZOD8wiz5YV9zktRxIL:-KZPDtenag9tMdJFDiKW:b-24hr6x) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/systems/controllers/zmp_planner.h#L57)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, I don't understand the second line here. Are you saying these are the same thing? The code seems just to implement the Cx+Du form. Please clarify.
</blockquote></details>

I added  a more detailed description at the top. 

---

*[drake/systems/controllers/zmp_planner.h, line 62 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZOJ8N_sNVp05IbcsNM:-KZP3Po6XAZtuuV6Nv1T:b-k93gqi) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/systems/controllers/zmp_planner.h#L62)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Don't put `inline` here. It is implied by putting the implementation in the class and redundant here. The compiler will decide whether the method is actually inlined regardless of your request here. (You do need `inline` if you put the declaration here but put the definition at the end of the header file.)
</blockquote></details>

Oh, that's good to know.

---

*[drake/systems/controllers/zmp_planner.h, line 68 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZOG_wJqzvXPwLsvMzf:-KZPDyyufjkjHZz8r8GU:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/systems/controllers/zmp_planner.h#L68)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

It would be useful to know when this method can be expected to return a valid result.
</blockquote></details>

Done.

---

*[drake/systems/controllers/zmp_planner.h, line 217 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZOHr1mr1W1vGbggfJl:-KZPE-ounjpDdzl-OxQR:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ac5b04944399c9c735e7cf49034d2015c417b314/drake/systems/controllers/zmp_planner.h#L217)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, this nice comment seems like it would be better placed in the public doxygen (and the ones below too).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-:-KZPGHiswewUrq02Q9ht:bgsir6n)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 1 of 10 files at r1, 2 of 6 files at r4, 1 of 5 files at r5, 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[drake/systems/controllers/zmp_planner.h, line 38 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZPUzAgNY7HQylzUo6K:-KZPUzAgNY7HQylzUo6L:b-frgbz) ([raw file](https://github.com/robotlocomotion/drake/blob/6fde43150f6a16728fcf1b21fd8c77fd461a9fcb/drake/systems/controllers/zmp_planner.h#L38)):*
> ```objc
>  * trajectory at time t.
>  * The value function (cost-to-go) is defined as
>  * V(x, t) = x_bar(tf)^T * S * x_bar(tf) + integral L(y, u, dt) dt,
> ```

Is value function the OPTIMAL cost-to-go? Namely we should add a min in the definition; besides, should we use L(y,u)dt instead of L(y,u,dt)dt?

I would suggest to write this equation using Latex also, so that we can use the integral symbol with tf as the upper-script, and 0 as the under-script.

---

*[drake/systems/controllers/zmp_planner.h, line 40 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZPVrPpuTxfU7aYsENr:-KZPVrPpuTxfU7aYsENs:bwxd678) ([raw file](https://github.com/robotlocomotion/drake/blob/6fde43150f6a16728fcf1b21fd8c77fd461a9fcb/drake/systems/controllers/zmp_planner.h#L40)):*
> ```objc
>  * V(x, t) = x_bar(tf)^T * S * x_bar(tf) + integral L(y, u, dt) dt,
>  * where tf is the last time in the desired ZMP trajectory,
>  * x_bar is defined as [com - y_d(tf); comd], S is the quadratic term from the
> ```

this is not y_d(tf), but com_d(tf)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4483)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Thanks for the changes, they look great. A few minor things remaining, PTAL.

---

Reviewed 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/controllers/zmp_planner.h, line 21 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZPnc3etxWCdbxj6Wa5:-KZPnc3etxWCdbxj6Wa6:b-70p7lg) ([raw file](https://github.com/robotlocomotion/drake/blob/6fde43150f6a16728fcf1b21fd8c77fd461a9fcb/drake/systems/controllers/zmp_planner.h#L21)):*
> ```objc
>  * The system dynamics for the X and Y directions are decoupled. We plan the
>  * XY motion together for convenience.
>  * For the X direction, the LIPM dynamics can is:
> ```

BTW, delete "can"?

---

*[drake/systems/controllers/zmp_planner.h, line 270 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZPnENqH2l01nzXmmoo:-KZPnENrgXdGKookZ3sm:b-y8spav) ([raw file](https://github.com/robotlocomotion/drake/blob/6fde43150f6a16728fcf1b21fd8c77fd461a9fcb/drake/systems/controllers/zmp_planner.h#L270)):*
> ```objc
>   // stationary or not in CheckStationaryEndPoint. This number is currently
>   // arbitrarily chosen.
>   double kStaitionaryThreshold = 1e-8;
> ```

Typo -- Staitionary -> Stationary.

---

*[drake/systems/controllers/zmp_planner.h, line 313 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZPob3Y6ul2Wc1KSn_6:-KZPob3Y6ul2Wc1KSn_7:b-1lm866) ([raw file](https://github.com/robotlocomotion/drake/blob/6fde43150f6a16728fcf1b21fd8c77fd461a9fcb/drake/systems/controllers/zmp_planner.h#L313)):*
> ```objc
>   ExponentialPlusPiecewisePolynomial<double> k2_;
> 
>   bool planned_;
> ```

Style guide requires initialization here: `bool planned_{false}`. Then you don't have to initialize it in the constructor.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-:-KZPoqINpnJA59mYMqlt:b9s41d3)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
@hongkai-dai I updated the block comments more, and this is what the doxygen spits out. 
Do we have a bib reference thing setup in drake now? How are we citing papers? 
![Screenshot from 2016-12-20 10:31:30.png](https://files.reviewableusercontent.io/images/reviewable/19416233/pfOvDHLm8hEMMLrMQN6gxXGr/Screenshot%20from%202016-12-20%2010%3A31%3A30.png)

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/controllers/zmp_planner.h, line 21 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZPnc3etxWCdbxj6Wa5:-KZRewqmM4C5o3eCiKlM:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6fde43150f6a16728fcf1b21fd8c77fd461a9fcb/drake/systems/controllers/zmp_planner.h#L21)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, delete "can"?
</blockquote></details>

Done.

---

*[drake/systems/controllers/zmp_planner.h, line 38 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZPUzAgNY7HQylzUo6K:-KZRf8_nKcdb-nEP9FpB:bsoqldq) ([raw file](https://github.com/robotlocomotion/drake/blob/6fde43150f6a16728fcf1b21fd8c77fd461a9fcb/drake/systems/controllers/zmp_planner.h#L38)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Is value function the OPTIMAL cost-to-go? Namely we should add a min in the definition; besides, should we use L(y,u)dt instead of L(y,u,dt)dt?

I would suggest to write this equation using Latex also, so that we can use the integral symbol with tf as the upper-script, and 0 as the under-script.
</blockquote></details>

opps. missed one important min.

---

*[drake/systems/controllers/zmp_planner.h, line 40 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZPVrPpuTxfU7aYsENr:-KZRfV98abxeZenmJ74V:b-10vh3x) ([raw file](https://github.com/robotlocomotion/drake/blob/6fde43150f6a16728fcf1b21fd8c77fd461a9fcb/drake/systems/controllers/zmp_planner.h#L40)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

this is not y_d(tf), but com_d(tf)?
</blockquote></details>

It's actually y_d(tf), see eq 5. There is no com_d, I suppose you can argue that's also com_d(tf) if we think it's stationary. 

---

*[drake/systems/controllers/zmp_planner.h, line 270 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZPnENqH2l01nzXmmoo:-KZRg85V3lIRr7BhpmKR:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6fde43150f6a16728fcf1b21fd8c77fd461a9fcb/drake/systems/controllers/zmp_planner.h#L270)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Typo -- Staitionary -> Stationary.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-:-KZRrKKsykq0gZfd3x9r:b-y09m0h)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: 9 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/controllers/zmp_planner.h, line 313 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-KZPob3Y6ul2Wc1KSn_6:-KZRspfIS81px0BMLRhK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6fde43150f6a16728fcf1b21fd8c77fd461a9fcb/drake/systems/controllers/zmp_planner.h#L313)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Style guide requires initialization here: `bool planned_{false}`. Then you don't have to initialize it in the constructor.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4483)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
That doxygen looks great! The header file is harder to read now but I think in this case the reader is just going to have to look at the doxygen. Platform review <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>.

---

Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-:-KZS7VZ5kXDuHv6m5I1o:b55fpfk)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
The doxygen looks AWESOME!!

One suggestion, we used both zero moment point and center of pressure. Could you mention they are the same thing? Also you may want to say `z is the CONSTANT CoM height`. <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-:-KZSEZNpoUupdrSFmhn9:b-1msaq3)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Thanks Hongkai, I added a sentence for const z, but I am going to leave the cop = zmp thing alone though. 

---

Review status: 10 of 11 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4483#-:-KZSNZM1qH9uwX-_EEm7:b-tsdeyh)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
@drake-jenkins-bot linux-clang-ninja-continuous-debug please
